tests/firedrake/adjoint/test_rf_eigensolver.py

```diff
@@ -0,0 +1,192 @@
+    new_restricted_control_variable,
+    interpolate_vars,
+)
+
+
+def generate_expressions():
+    """
+    Returns a list of three expressions.
+    """
+    mesh = UnitSquareMesh(50, 50, quadrilateral=False)
+    x, y = SpatialCoordinate(mesh)
+    return [2 * x, exp(-x - y), sin(pi * x * y)]
+
+
+@pytest.mark.parametrize("expression", generate_expressions())
+def test_compare_eigensolvers_helmholtz(expression):
+    """Test TLM vs assembled form for Helmholtz eigenproblem."""
+    n = 50
+    num_eigenvalues = 10
+
+    mesh = UnitSquareMesh(n, n, quadrilateral=False)
+    x, y = SpatialCoordinate(mesh)
+    V = FunctionSpace(mesh, "CG", 1)
+    bc = DirichletBC(V, 0.0, "on_boundary")
+
+    u = TrialFunction(V)
+    v = TestFunction(V)
+    a = (inner(grad(u), grad(v)) + u * v) * dx
+
+    # Form based Ax = cMx
+    problem_form = LinearEigenproblem(a, bcs=bc, restrict=True)
+    solver_form = LinearEigensolver(problem_form, n_evals=num_eigenvalues)
+    solver_form.solve()
+    nconv = solver_form.es.getConverged()
+
+    evals_form = [
+        solver_form.eigenvalue(i) for i in range(0, min(num_eigenvalues, nconv))
+    ]
+
+    # ReducedFunctional-based, Jx = 1/cx
+    continue_annotation()
+    f = Function(V)
+    f.interpolate(2 * x)
+    control = Control(f)
+    u_1 = Function(V)
+    v = TestFunction(V)
+    F = (inner(grad(u_1), grad(v)) + u_1 * v - f * v) * dx
+    solve(F == 0, u_1, bcs=bc)
+    J = ReducedFunctional(u_1, controls=[control])
+    pause_annotation()
+
+    problem_rf = RFEigenproblem(J, bcs=bc, apply_riesz=False, restrict=True, identity=True)
+    solver_rf = RFEigensolver(problem_rf, n_evals=num_eigenvalues)
+    solver_rf.es.setTarget(1)
+    solver_rf.solve()
+    nconv = solver_rf.es.getConverged()
+
+    evals_rf = [
+        1 / solver_rf.eigenvalue(i) for i in range(0, min(num_eigenvalues, nconv))
+    ]
+
+    for ev1, ev2 in zip(evals_form, evals_rf):
+        assert abs(ev1 - ev2) < 1e-6
+
+    set_working_tape(Tape())
+
+
+@pytest.mark.parametrize("restrict", [False])
+@pytest.mark.parametrize("riesz", ["L2"])
+def test_apply_riesz_param(restrict, riesz):
+    """Test implicit matrix for Helmholtz eigenproblem."""
+    n = 50
+    mesh = UnitSquareMesh(n, n, quadrilateral=False)
+    x, y = SpatialCoordinate(mesh)
+    V = FunctionSpace(mesh, "CG", 1)
+    bc = DirichletBC(V, 0.0, "on_boundary")
+
+    # Tape the solve
+    continue_annotation()
+    f = Function(V)
+    f.interpolate(2 * x * y)
+    control = Control(f)
+    u_1 = Function(V)
+    v = TestFunction(V)
+    F = (inner(grad(u_1), grad(v)) + u_1 * v - f * v) * dx
+    solve(F == 0, u_1, bcs=bc)
+    J = ReducedFunctional(u_1, controls=[control])
+    pause_annotation()
+
+    # Check adjoint action
+    problem_rf_true = RFEigenproblem(
+        J, bcs=bc, apply_riesz=True, restrict=restrict, action="adjoint"
+    )
+    problem_rf_false = RFEigenproblem(
+        J, bcs=bc, apply_riesz=False, restrict=restrict, action="adjoint"
+    )
```

Comment:
Seeing as we're specifically testing the implicit matrix here, I think it'd be simpler to just construct them explicitly rather than going via the RFEigenProblem.